System.BadImageFormatException when reflecting on BCL shipped with Mono

Mono throws exception when methods in the BCL are queried for custom attributes.

## Steps to Reproduce

1. Run test case below

```
using System;
using System.Reflection;
using System.Linq;
using System.IO;

internal class Program
{
    public static int Main(string[] args)
    {
        foreach (Type type in typeof(System.Linq.Enumerable).Assembly.GetTypes())
        {
            foreach (MemberInfo method in type.GetMembers(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic))
            {
                try
                {
                    method.GetCustomAttributes<DummyAttribute>();
                }
                catch (BadImageFormatException)
                {
                    Console.WriteLine($"BadImageFormatException {type}:{method}");
                    throw;
                }
            }
        }
        return 0;
    }
}

class DummyAttribute : Attribute
{

}

```

### Current Behavior

```
System.Linq.Parallel.GroupJoinQueryOperator`4[TLeftInput,TRightInput,TKey,TOutput]:System.Linq.Parallel.QueryResults`1[TOutput] Open(System.Linq.Parallel.QuerySettings, Boolean)

Unhandled Exception:
System.BadImageFormatException: VAR 3 (TOutput) cannot be expanded in this context with 3 instantiations
  at (wrapper managed-to-native) System.Reflection.RuntimeMethodInfo.get_base_method(System.Reflection.RuntimeMethodInfo,bool)
  at System.Reflection.RuntimeMethodInfo.GetBaseMethod () [0x00000] in <b814b509d4ad406fb40c6c93e38929e7>:0 
  at System.MonoCustomAttrs.GetBase (System.Reflection.ICustomAttributeProvider obj) [0x000ad] in <b814b509d4ad406fb40c6c93e38929e7>:0 
  at System.MonoCustomAttrs.GetCustomAttributes (System.Reflection.ICustomAttributeProvider obj, System.Type attributeType, System.Boolean inherit) [0x000b5] in <b814b509d4ad406fb40c6c93e38929e7>:0 
  at System.Reflection.RuntimeMethodInfo.GetCustomAttributes (System.Type attributeType, System.Boolean inherit) [0x00000] in <b814b509d4ad406fb40c6c93e38929e7>:0 
  at System.Attribute.GetCustomAttributes (System.Reflection.MemberInfo element, System.Type type, System.Boolean inherit) [0x00088] in <b814b509d4ad406fb40c6c93e38929e7>:0 
  at System.Attribute.GetCustomAttributes (System.Reflection.MemberInfo element, System.Type type) [0x00000] in <b814b509d4ad406fb40c6c93e38929e7>:0 
  at System.Reflection.CustomAttributeExtensions.GetCustomAttributes (System.Reflection.MemberInfo element, System.Type attributeType) [0x00000] in <b814b509d4ad406fb40c6c93e38929e7>:0 
  at System.Reflection.CustomAttributeExtensions.GetCustomAttributes[T] (System.Reflection.MemberInfo element) [0x00000] in <b814b509d4ad406fb40c6c93e38929e7>:0 
  at Program.Main (System.String[] args) [0x00057] in <a28b7bf7d0b34731b1ba0283bc2d394c>:0 
[ERROR] FATAL UNHANDLED EXCEPTION: System.BadImageFormatException: VAR 3 (TOutput) cannot be expanded in this context with 3 instantiations
  at (wrapper managed-to-native) System.Reflection.RuntimeMethodInfo.get_base_method(System.Reflection.RuntimeMethodInfo,bool)
  at System.Reflection.RuntimeMethodInfo.GetBaseMethod () [0x00000] in <b814b509d4ad406fb40c6c93e38929e7>:0 
  at System.MonoCustomAttrs.GetBase (System.Reflection.ICustomAttributeProvider obj) [0x000ad] in <b814b509d4ad406fb40c6c93e38929e7>:0 
  at System.MonoCustomAttrs.GetCustomAttributes (System.Reflection.ICustomAttributeProvider obj, System.Type attributeType, System.Boolean inherit) [0x000b5] in <b814b509d4ad406fb40c6c93e38929e7>:0 
  at System.Reflection.RuntimeMethodInfo.GetCustomAttributes (System.Type attributeType, System.Boolean inherit) [0x00000] in <b814b509d4ad406fb40c6c93e38929e7>:0 
  at System.Attribute.GetCustomAttributes (System.Reflection.MemberInfo element, System.Type type, System.Boolean inherit) [0x00088] in <b814b509d4ad406fb40c6c93e38929e7>:0 
  at System.Attribute.GetCustomAttributes (System.Reflection.MemberInfo element, System.Type type) [0x00000] in <b814b509d4ad406fb40c6c93e38929e7>:0 
  at System.Reflection.CustomAttributeExtensions.GetCustomAttributes (System.Reflection.MemberInfo element, System.Type attributeType) [0x00000] in <b814b509d4ad406fb40c6c93e38929e7>:0 
  at System.Reflection.CustomAttributeExtensions.GetCustomAttributes[T] (System.Reflection.MemberInfo element) [0x00000] in <b814b509d4ad406fb40c6c93e38929e7>:0 
  at Program.Main (System.String[] args) [0x00057] in <a28b7bf7d0b34731b1ba0283bc2d394c>:0 
```

### Expected Behavior

No Exception

## On which platforms did you notice this

[X ] macOS
[ ] Linux
[ X] Windows

**Version Used**:

```
Mono JIT compiler version 6.4.0.198 (2019-06/fe64a4765e6 Wed Sep 18 14:14:51 EDT 2019)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           
	SIGSEGV:       altstack
	Notification:  kqueue
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug 
	Interpreter:   yes
	LLVM:          yes(610)
	Suspend:       hybrid
	GC:            sgen (concurrent by default)
```

Corresponds to Unity Issue: https://issuetracker.unity3d.com/issues/badimageformatexception-is-thrown-when-calling-getcustomattributes-on-certain-memberinfo-instances

--- a/mcs/class/corlib/Test/System/AttributeTest.cs
+++ b/mcs/class/corlib/Test/System/AttributeTest.cs
@@ -1252,6 +1252,27 @@ namespace MonoTests.System
 			Assert.AreEqual (1, attrs.Length);	
 		}
 
+		[Test]
+		public void GetCustomAttributesOnOpenGenericMethodOverride ()
+		{
+			// Regression test for https://github.com/mono/mono/issues/17278
+			/* The issue is in how we pass along the generic
+			 * instantiation.  So there are two parts here: we
+			 * start with a method on a GTD which has to decompose
+			 * into the GTD and it's instantiation with its own
+			 * generic parameters, and then we go up to its parent
+			 * and grandparent inflating and decomposing the
+			 * resulting instantiation. The instantiations keep
+			 * passing the last parameter as the number of
+			 * parameters shrinks so that the gparam index from the
+			 * original type is now out of bounds for the parent
+			 * and grandparent.  If we mess up, we'll get a
+			 * BadImageFormatException from the runtime. */
+			var m = typeof (Three<,,>).GetMethod ("Foo");
+			var attrs = Attribute.GetCustomAttributes(m, typeof(MyCAttr), true);
+			Assert.AreEqual (1, attrs.Length);
+		}
+
 		class EvtBase
 		{
 			public virtual event EventHandler Event {add {} remove {}}
@@ -1266,6 +1287,20 @@ namespace MonoTests.System
 		class EvtChild : EvtOverride
 		{
 			public override event EventHandler Event {add {} remove {}}
+		}
+
+		abstract class One<A> {
+			[MyCAttr]
+			public abstract A Foo ();
+		}
+
+		abstract class Two<X,Y> : One<Y> {
+		}
+
+		class Three<P,Q,R> : Two<Q,R> {
+			public override R Foo () {
+				throw new Exception ("Doesn't matter");
+			}
 		}
 
 		[Test] //Regression test for #662867

--- a/mono/metadata/class.c
+++ b/mono/metadata/class.c
@@ -6253,7 +6253,17 @@ mono_method_get_base_method (MonoMethod *method, gboolean definition, MonoError 
 		return method;
 
 	klass = method->klass;
-	if (mono_class_is_ginst (klass)) {
+	if (mono_class_is_gtd (klass)) {
+		/* If we get a GTD like Foo`2 replace look instead at its instantiation with its own generic params: Foo`2<!0, !1>. */
+		/* In particular we want generic_inst to be initialized to <!0,
+		 * !1> so that we can inflate parent classes correctly as we go
+		 * up the class hierarchy. */
+		MonoType *ty = mono_class_gtd_get_canonical_inst (klass);
+		g_assert (ty->type == MONO_TYPE_GENERICINST);
+		MonoGenericClass *gklass = ty->data.generic_class;
+		generic_inst = mono_generic_class_get_context (gklass);
+		klass = gklass->container_class;
+	} else if (mono_class_is_ginst (klass)) {
 		generic_inst = mono_class_get_context (klass);
 		klass = mono_class_get_generic_class (klass)->container_class;
 	}
@@ -6301,6 +6311,10 @@ retry:
 			generic_inst = parent_inst;
 		}
 	} else {
+		/* When we get here, possibly after a retry, if generic_inst is
+		 * set, then the class is must be a gtd */
+		g_assert (generic_inst == NULL || mono_class_is_gtd (klass));
+
 		klass = m_class_get_parent (klass);
 		if (!klass)
 			return method;
@@ -6320,6 +6334,7 @@ retry:
 	if (generic_inst) {
 		klass = mono_class_inflate_generic_class_checked (klass, generic_inst, error);
 		return_val_if_nok (error, NULL);
+		generic_inst = NULL;
 	}
 
 	if (klass == method->klass)

